### PR TITLE
Migrate release workflow to `cargo-dist`

### PR DESCRIPTION
**Pull request by charliermarsh**
## Summary

This PR migrates our release workflow to [`cargo-dist`](https://github.com/axodotdev/cargo-dist). The primary motivation here is that we want to ship dedicated installers for Ruff that work across platforms, and `cargo-dist` gives us those installers out-of-the-box. The secondary motivation is that `cargo-dist` formalizes some of the patterns that we've built up over time in our own release process.

At a high level:

- The `release.yml` file is generated by `cargo-dist` with `cargo dist generate`. It doesn't contain any modifications vis-a-vis the generated file. (If it's edited out of band from generation, the release fails.)
- Our customizations are inserted as custom steps within the `cargo-dist` workflow. Specifically, `build-binaries` builds the wheels and packages them into binaries (as on `main`), while `build-docker.yml` builds the Docker image. `publish-pypi.yml` publishes the wheels to PyPI. This is effectively our `release.yaml` (on `main`), broken down into individual workflows rather than steps within a single workflow.

### Changes from `main`

The workflow is _nearly_ unchanged. We kick off a release manually via the GitHub Action by providing a tag. If the tag doesn't match the `Cargo.toml`, the release fails. If the tag matches an already-existing release, the release fails.

The release proceeds by (in order):

0. Doing some upfront validation via `cargo-dist`.
1. Creating the wheels and archives.
2. Building and pushing the Docker image.
3. Publishing to PyPI (if it's not a "dry run").
4. Creating the GitHub Release (if it's not a "dry run").
5. Notifying `ruff-pre-commit` (if it's not a "dry run").

There are a few changes in the workflow as compared to `main`:

- **We no longer validate the SHA** (just the tag). It's not an input to the job. The Axo team is considering whether / how to support this.
- **Releases are now published directly** (rather than as draft). Again, the Axo team is considering whether / how to support this. The downside of drafts is that the URLs aren't stable, so the installers don't work _as long as the release is in draft_. This is fine for our workflow. It seems like the Axo team will add it.
- Releases already contain the latest entry from the changelog (we don't need to copy it over). This "Just Works", which is nice, though we'll still want to edit them to add contributors.

There are also a few **breaking changes** for consumers of the binaries:

- **We no longer include the version tag in the file name**. This enables users to install via `/latest` URLs on GitHub, and is part of the cargo-dist paradigm.
- **Archives now include an extra level of nesting,** which you can remove with `--strip-components=1` when untarring.

Here's an example release that I created -- I omitted all the artifacts since I was just testing a workflow, so none of the installers or links work, but it gives you a sense for what the release looks like: https://github.com/charliermarsh/cargodisttest/releases/tag/0.1.13.

### Test Plan

I ran a successful release to completion last night, and installed Ruff via the installer:

Screenshot 2024-01-17 at 12 12 53 AM

Screenshot 2024-01-17 at 12 12 48 AM

The piece I'm least confident about is the Docker push. We build the image, but the push fails in my test repo since I haven't wired up the credentials.

### CODE REVIEW

File: .github/workflows/build-binaries.yml [hunk 1]
```diff
@@ -86,13 +88,18 @@ jobs:
...
-          ARCHIVE_FILE=ruff-${{ inputs.tag }}-x86_64-apple-darwin.tar.gz
-          tar czvf $ARCHIVE_FILE -C target/x86_64-apple-darwin/release ruff
+          TARGET=x86_64-apple-darwin
+          ARCHIVE_NAME=ruff-$TARGET
+          ARCHIVE_FILE=$ARCHIVE_NAME.tar.gz
+
+          mkdir -p $ARCHIVE_NAME
+          cp target/$TARGET/release/ruff $ARCHIVE_NAME/ruff
+          tar czvf $ARCHIVE_FILE $ARCHIVE_NAME
           shasum -a 256 $ARCHIVE_FILE > $ARCHIVE_FILE.sha256
```
Comment (charliermarsh) [inline .github/workflows/build-binaries.yml:98]: One import piece here is that per @Gankra the installers expect that the binary is under a dummy directory of the same name, so we copy the binary into a dummy directory of the same name prior to archiving.

File: .github/workflows/build-docker.yml [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+      - name: "Build and push Docker image"
+        uses: docker/build-push-action@v5
+        with:
+          context: .
+          platforms: linux/amd64,linux/arm64
+          # Reuse the builder
+          cache-from: type=gha
+          cache-to: type=gha,mode=max
+          push: ${{ !fromJson(inputs.plan).announcement_tag_is_implicit }}
```
Comment (charliermarsh) [inline .github/workflows/build-docker.yml:62]: `announcement_tag_is_implicit` indicates "dry run".

File: Cargo.toml [hunk 3]
```diff
@@ -175,3 +175,38 @@ opt-level = 1
...
+# CI backends to support
+ci = ["github"]
+# The installers to generate for each app
+installers = ["shell", "powershell"]
+# The archive format to use for windows builds (defaults .zip)
+windows-archive = ".zip"
+# The archive format to use for non-windows builds (defaults .tar.xz)
+unix-archive = ".tar.gz"
+# Target platforms to build apps for (Rust target-triple syntax)
+targets = ["aarch64-unknown-linux-gnu", "x86_64-unknown-linux-gnu", "i686-unknown-linux-gnu", "aarch64-apple-darwin", "x86_64-apple-darwin", "aarch64-unknown-linux-musl", "x86_64-unknown-linux-musl", "i686-unknown-linux-musl", "aarch64-pc-windows-msvc", "x86_64-pc-windows-msvc", "i686-pc-windows-msvc", "armv7-unknown-linux-gnueabihf", "powerpc64-unknown-linux-gnu", "powerpc64le-unknown-linux-gnu", "s390x-unknown-linux-gnu"]
```
Comment (zanieb) [inline Cargo.toml:196]: Could we split these onto multiple lines?

File: Cargo.toml [hunk 4]
```diff
@@ -175,3 +175,38 @@ opt-level = 1
...
+# Whether to auto-include files like READMEs and CHANGELOGs (default true)
+auto-includes = false
```
Comment (zanieb) [inline Cargo.toml:198]: Why false?

File: .github/workflows/release.yml [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+      - name: Install cargo-dist
+        # we specify bash to get pipefail; it guards against the `curl` command
+        # failing. otherwise `sh` won't catch that `curl` returned non-0
+        shell: bash
+        run: "curl --proto '=https' --tlsv1.2 -LsSf https://github.com/axodotdev/cargo-dist/releases/download/v0.7.3-prerelease.3/cargo-dist-installer.sh | sh"
+      # sure would be cool if github gave us proper conditionals...
+      # so here's a doubly-nested ternary-via-truthiness to try to provide the best possible
+      # functionality based on whether this is a pull_request, and whether it's from a fork.
+      # (PRs run on the *source* but secrets are usually on the *target* -- that's *good*
+      # but also really annoying to build CI around when it needs secrets to work right.)
```
Comment (zanieb) [inline .github/workflows/release.yml:73]: Not for review, but I find this confusing and it doesn't make it clear to me what the behaviors are on pull requests and forks.

File: .github/workflows/build-binaries.yml [hunk 6]
```diff
@@ -223,14 +235,22 @@ jobs:
           name: wheels
           path: dist
       - name: "Archive binary"
+        shell: bash
         run: |
-          ARCHIVE_FILE=ruff-${{ inputs.tag }}-${{ matrix.target }}.tar.gz
-          tar czvf $ARCHIVE_FILE -C target/${{ matrix.target }}/release ruff
+          set -euo pipefail
```
Comment (zanieb) [inline .github/workflows/build-binaries.yml:240]: Why do this here but not above?

File: .github/workflows/build-binaries.yml [hunk 7]
```diff
@@ -312,7 +340,7 @@ jobs:
     steps:
       - uses: actions/checkout@v4
         with:
-          ref: ${{ inputs.sha }}
+          submodules: recursive
```
Comment (zanieb) [inline .github/workflows/build-binaries.yml:343]: We have submodules?

**Review by zanieb — APPROVED** [review]

File: .github/workflows/build-binaries.yml [hunk 8]
```diff
@@ -1,21 +1,23 @@
...
+      plan:
+        required: true
         type: string
   pull_request:
     paths:
       # When we change pyproject.toml, we want to ensure that the maturin builds still work
       - pyproject.toml
       # And when we change this workflow itself...
-      - .github/workflows/release.yaml
+      - .github/workflows/build.yaml
```
Comment (konstin) [inline .github/workflows/build-binaries.yml:20]: There's no file with that name, maybe when any workflow changes, or just `.github/workflows/build-binaries.yaml`

```suggestion
      - .github/workflows/*.yaml
```

File: .github/workflows/build-binaries.yml (hunk 1 above)
Comment (konstin) [reply .github/workflows/build-binaries.yml:98]: Could you make this a code comment?

File: .github/workflows/build-docker.yml [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+          username: ${{ github.repository_owner }}
+          password: ${{ secrets.GITHUB_TOKEN }}
+
+      - name: Extract metadata (tags, labels) for Docker
+        id: meta
+        uses: docker/metadata-action@v5
+        with:
+          images: ghcr.io/astral-sh/ruff
+
+      - name: Check tag consistency
```
Comment (konstin) [inline .github/workflows/build-docker.yml:42]: Is this one intentionally only in the docker build?

**Review by konstin — APPROVED** [review]

**Comment by charliermarsh** [comment]
I'm planning to ship this as part of Ruff v0.4.0 (i.e., in the next few weeks) given that it changes the archive format.

**Comment by charliermarsh** [comment]
I will merge this today.

**Pull request merged**